x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java

```diff
@@ -0,0 +1,552 @@
+            logger.trace("received response for node {}", nodeId);
+            nodesResponses.set(nodeIndex, response);
+            if (expectedOps == counterOps.incrementAndGet()) {
+                finishHim(true);
+            } else {
+                checkForEarlyFinish();
+            }
+        }
+
+        void onNoOperation(String nodeId) {
+            if (expectedOps == counterOps.incrementAndGet()) {
+                finishHim(true);
+            }
+        }
+
+        // Can be called multiple times - either for early time-outs or for fully-completed collections.
+        protected synchronized void finishHim(boolean complete) {
+            if (listener == null) {
+                return;
+            }
+            try {
+                listener.onResponse(newResponse(request, nodesResponses, complete, nodeBundles));
+            } catch (Exception e) {
+                listener.onFailure(e);
+            } finally {
+                listener = null;
+            }
+        }
+    }
+
+    class NodeTransportHandler implements TransportRequestHandler<NodeTermEnumRequest> {
+
+        @Override
+        public void messageReceived(NodeTermEnumRequest request, TransportChannel channel, Task task) throws Exception {
+            asyncNodeOperation(request, task, ActionListener.wrap(channel::sendResponse, e -> {
+                try {
+                    channel.sendResponse(e);
+                } catch (Exception e1) {
+                    logger.warn(
+                        () -> new ParameterizedMessage(
+                            "Failed to send error response for action [{}] and request [{}]",
+                            actionName,
+                            request
+                        ),
+                        e1
+                    );
+                }
+            }));
+        }
+    }
+
+    private void asyncNodeOperation(NodeTermEnumRequest request, Task task, ActionListener<NodeTermEnumResponse> listener)
+        throws IOException {
+        // DLS/FLS check copied from ResizeRequestInterceptor - check permissions and
+        // any index_filter canMatch checks on network thread before allocating work
+        ThreadContext threadContext = transportService.getThreadPool().getThreadContext();
+        final XPackLicenseState frozenLicenseState = licenseState.copyCurrentLicenseState();
+        for (ShardId shardId : request.shardIds().toArray(new ShardId[0])) {
+            if (canAccess(shardId.getIndexName(), request.field(), frozenLicenseState, threadContext) == false || canMatchShard(
+                shardId,
+                request
+            ) == false) {
+                // Permission denied or can't match, remove shardID from request
+                request.remove(shardId);
+            }
+        }
+        if (request.shardIds().size() == 0) {
+            listener.onResponse(new NodeTermEnumResponse(request.nodeId(), Collections.emptyList(), null, true));
+        } else {
```

Comment:
We could also use the search thread pool if the queue is empty. Happy to do it in a follow up.